src/popover/utils/positions.ts

```diff
@@ -228,55 +227,74 @@ export function calculatePosition({
   viewport,
   // the popover is only bound by the viewport if it is rendered in a portal
   renderWithPortal,
+  allowVerticalOverflow,
 }: {
   preferredPosition: PopoverProps.Position;
   fixedInternalPosition?: InternalPosition;
-  trigger: BoundingOffset;
-  arrow: BoundingBox;
-  body: BoundingBox;
-  container: BoundingOffset;
-  viewport: BoundingOffset;
+  trigger: BoundingBox;
+  arrow: Dimensions;
+  body: Dimensions;
+  container: BoundingBox;
+  viewport: BoundingBox;
   // the popover is only bound by the viewport if it is rendered in a portal
   renderWithPortal?: boolean;
+  allowVerticalOverflow?: boolean;
 }): CalculatedPosition {
-  let bestPositionOutsideViewport: CalculatedPosition | null = null;
-  let largestArea = 0;
+  let bestOption: CandidatePosition | null = null;
 
   // If a fixed internal position is passed, only consider this one.
   const preferredInternalPositions = fixedInternalPosition
     ? [fixedInternalPosition]
     : PRIORITY_MAPPING[preferredPosition];
 
-  // Attempt to position the popover based on the priority list for this position,
-  // trying to fit it inside the container and inside the viewport.
+  // Attempt to position the popover based on the priority list for this position.
   for (const internalPosition of preferredInternalPositions) {
-    const boundingOffset = RECTANGLE_CALCULATIONS[internalPosition]({ body, trigger, arrow });
-    const fitsInContainer = renderWithPortal || canRectFit(boundingOffset, container);
-    const fitsInViewport = canRectFit(boundingOffset, viewport);
-    if (fitsInContainer && fitsInViewport) {
-      return { internalPosition, boundingOffset };
-    }
-    const boundingRectangles = [boundingOffset, viewport];
-    if (!renderWithPortal) {
-      boundingRectangles.push(container);
-    }
-    const availableArea = intersectRectangles(boundingRectangles);
-    if (availableArea && availableArea > largestArea) {
-      bestPositionOutsideViewport = { internalPosition, boundingOffset };
-      largestArea = availableArea;
+    const boundingBox = RECTANGLE_CALCULATIONS[internalPosition]({ body, trigger, arrow });
+    const visibleArea = renderWithPortal
+      ? getIntersection([boundingBox, viewport])
+      : getIntersection([boundingBox, viewport, container]);
+
+    const fitsWithoutOverflow = visibleArea && visibleArea.width === body.width && visibleArea.height === body.height;
+
+    if (fitsWithoutOverflow) {
+      return { internalPosition, boundingBox };
     }
+
+    const newOption = { boundingBox, internalPosition, visibleArea };
+    bestOption = getBestOption(newOption, bestOption);
   }
 
   // Use best possible placement.
-  const internalPosition = bestPositionOutsideViewport?.internalPosition || 'right-top';
+  const internalPosition = bestOption?.internalPosition || 'right-top';
```

Comment:
Is this a part of the issue, that by default we fallback to `right-top` position, which is hard to align on small screens?

Should we change it to something more reasonable?

Reply:
This should not be relevant because the only case where this value would be used would be if _all positioning options_ would be completely off screen (not a single pixel inside the viewport). We could change this, but it is not the cause of the issue.